Source/DafnyLanguageServer.Test/DafnyLanguageServerTestBase.cs

```diff
@@ -60,30 +59,47 @@ ensures true
       this.output = new WriterFromOutputHelper(output);
     }
 
-    protected virtual IServiceCollection ServerOptionsAction(LanguageServerOptions serverOptions) {
-      return serverOptions.Services.AddSingleton<IProgramVerifier>(serviceProvider => new SlowVerifier(
-        serviceProvider.GetRequiredService<ILogger<DafnyProgramVerifier>>()
-      ));
+    protected virtual void ServerOptionsAction(LanguageServerOptions serverOptions) {
     }
 
-    protected virtual async Task<ILanguageClient> InitializeClient(
-      Action<LanguageClientOptions> clientOptionsAction = null,
-      Action<DafnyOptions> modifyOptions = null) {
-      var dafnyOptions = DafnyOptions.Create(output);
-      modifyOptions?.Invoke(dafnyOptions);
+    protected Task<(ILanguageClient client, ILanguageServer server)> Initialize(Action<LanguageClientOptions> clientOptionsAction, Action<DafnyOptions> serverOptionsAction) {
+      /*
+       * I would rather use LanguageServerOptions.RegisterForDisposal, but it doesn't seem to work
+       * Alternatively one can do Disposable.Add((IDisposable)Server.Services), but we've seen many
+       * ObjectDisposedExceptions in tests that might relate to this, so let's be more specific and only dispose
+       * IProjectDatabase
+       */
+      Disposable.Add(new AnonymousDisposable(() => {
+        var database = Server.Services.GetRequiredService<IProjectDatabase>();
+        database.Dispose();
+      }));
+      return base.Initialize(clientOptionsAction, GetServerOptionsAction(serverOptionsAction));
+    }
 
-      void NewServerOptionsAction(LanguageServerOptions options) {
-        ApplyDefaultOptionValues(dafnyOptions);
+    private sealed class AnonymousDisposable : IDisposable {
+      private Action action;
 
-        ServerCommand.ConfigureDafnyOptionsForServer(dafnyOptions);
-        options.Services.AddSingleton(dafnyOptions);
-        ServerOptionsAction(options);
+      public AnonymousDisposable(Action action) {
+        this.action = action;
       }
 
-      var client = CreateClient(clientOptionsAction, NewServerOptionsAction);
-      await client.Initialize(CancellationToken).ConfigureAwait(false);
+      public void Dispose() => Interlocked.Exchange(ref action, null)?.Invoke();
```

Comment:
What is Interlocked.Exchange doing?

Reply:
Making sure the dispose action is only run once. Don't think it's relevant here but good defensive programming.